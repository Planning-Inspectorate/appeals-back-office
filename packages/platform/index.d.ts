import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
