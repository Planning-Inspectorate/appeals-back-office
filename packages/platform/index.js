export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
