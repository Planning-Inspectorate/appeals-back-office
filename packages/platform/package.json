{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
