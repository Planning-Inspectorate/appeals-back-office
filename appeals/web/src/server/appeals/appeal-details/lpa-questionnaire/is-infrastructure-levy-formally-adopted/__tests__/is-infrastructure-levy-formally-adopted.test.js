// @ts-nocheck
import { parseHtml } from '@pins/platform';
import supertest from 'supertest';
import {
	appealData,
	lpaQuestionnaireDataNotValidated
} from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import nock from 'nock';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';

describe('is-infrastructure-levy-formally-adopted', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	describe('GET /is-infrastructure-levy-formally-adopted/change', () => {
		it('should render the change levy formally adopted page with "Formally adopted" radio option checked if isInfrastructureLevyFormallyAdopted is true', async () => {
			nock('http://test/')
				.get(`/appeals/1/lpa-questionnaires/${appealData.lpaQuestionnaireId}`)
				.reply(200, {
					...lpaQuestionnaireDataNotValidated,
					isInfrastructureLevyFormallyAdopted: true
				});

			const response = await request.get(
				`${baseUrl}/1/lpa-questionnaire/${appealData.lpaQuestionnaireId}/is-infrastructure-levy-formally-adopted/change`
			);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change whether levy formally adopted</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="isInfrastructureLevyFormallyAdoptedRadio" type="radio" value="yes" checked>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'<label class="govuk-label govuk-radios__label" for="isInfrastructureLevyFormallyAdoptedRadio"> Formally adopted</label>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="isInfrastructureLevyFormallyAdoptedRadio" type="radio" value="no">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'<label class="govuk-label govuk-radios__label" for="isInfrastructureLevyFormallyAdoptedRadio-2"> Not formally adopted</label>'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render the change levy formally adopted page with "Not formally adopted" radio option checked if isInfrastructureLevyFormallyAdopted is false', async () => {
			nock('http://test/')
				.get(`/appeals/1/lpa-questionnaires/${appealData.lpaQuestionnaireId}`)
				.reply(200, {
					...lpaQuestionnaireDataNotValidated,
					isInfrastructureLevyFormallyAdopted: false
				});

			const response = await request.get(
				`${baseUrl}/1/lpa-questionnaire/${appealData.lpaQuestionnaireId}/is-infrastructure-levy-formally-adopted/change`
			);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change whether levy formally adopted</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="isInfrastructureLevyFormallyAdoptedRadio" type="radio" value="yes">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'<label class="govuk-label govuk-radios__label" for="isInfrastructureLevyFormallyAdoptedRadio"> Formally adopted</label>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="isInfrastructureLevyFormallyAdoptedRadio" type="radio" value="no" checked>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'<label class="govuk-label govuk-radios__label" for="isInfrastructureLevyFormallyAdoptedRadio-2"> Not formally adopted</label>'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});
	});

	describe('POST /is-infrastructure-levy-formally-adopted/change', () => {
		const validValues = ['yes', 'no'];

		for (const validValue of validValues) {
			it(`should call LPA questionnaires PATCH endpoint and redirect to the LPA questionnaire page if "${validValue}" is selected`, async () => {
				const mockLPAQPatchEndpoint = nock('http://test/')
					.patch(`/appeals/1/lpa-questionnaires/${appealData.lpaQuestionnaireId}`)
					.reply(200, {});

				const response = await request
					.post(
						`${baseUrl}/1/lpa-questionnaire/${appealData.lpaQuestionnaireId}/is-infrastructure-levy-formally-adopted/change`
					)
					.send({
						isInfrastructureLevyFormallyAdoptedRadio: validValue
					});

				expect(mockLPAQPatchEndpoint.isDone()).toBe(true);
				expect(response.statusCode).toBe(302);
				expect(response.text).toBe(
					`Found. Redirecting to /appeals-service/appeal-details/1/lpa-questionnaire/${appealData.lpaQuestionnaireId}`
				);
			});
		}
	});
});
