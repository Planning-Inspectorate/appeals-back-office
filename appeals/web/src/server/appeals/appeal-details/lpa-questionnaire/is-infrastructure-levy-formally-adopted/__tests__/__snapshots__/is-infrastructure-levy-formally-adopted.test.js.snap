// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is-infrastructure-levy-formally-adopted GET /is-infrastructure-levy-formally-adopted/change should render the change levy formally adopted page with "Formally adopted" radio option checked if isInfrastructureLevyFormallyAdopted is true 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change whether levy formally adopted</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="isInfrastructureLevyFormallyAdoptedRadio"
                            name="isInfrastructureLevyFormallyAdoptedRadio" type="radio" value="yes"
                            checked>
                            <label class="govuk-label govuk-radios__label" for="isInfrastructureLevyFormallyAdoptedRadio">Formally adopted</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="isInfrastructureLevyFormallyAdoptedRadio-2"
                            name="isInfrastructureLevyFormallyAdoptedRadio" type="radio" value="no">
                            <label class="govuk-label govuk-radios__label" for="isInfrastructureLevyFormallyAdoptedRadio-2">Not formally adopted</label>
                        </div>
                    </div>
                </div>
                <button type="submit" data-prevent-double-click="true" class="govuk-button"
                data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`is-infrastructure-levy-formally-adopted GET /is-infrastructure-levy-formally-adopted/change should render the change levy formally adopted page with "Not formally adopted" radio option checked if isInfrastructureLevyFormallyAdopted is false 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change whether levy formally adopted</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="isInfrastructureLevyFormallyAdoptedRadio"
                            name="isInfrastructureLevyFormallyAdoptedRadio" type="radio" value="yes">
                            <label class="govuk-label govuk-radios__label" for="isInfrastructureLevyFormallyAdoptedRadio">Formally adopted</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="isInfrastructureLevyFormallyAdoptedRadio-2"
                            name="isInfrastructureLevyFormallyAdoptedRadio" type="radio" value="no"
                            checked>
                            <label class="govuk-label govuk-radios__label" for="isInfrastructureLevyFormallyAdoptedRadio-2">Not formally adopted</label>
                        </div>
                    </div>
                </div>
                <button type="submit" data-prevent-double-click="true" class="govuk-button"
                data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`is-infrastructure-levy-formally-adopted POST /is-infrastructure-levy-formally-adopted/change should re-render the change levy formally adopted page with a validation error if a radio option was not selected 1`] = `"<div class="govuk-error-summary" data-module="govuk-error-summary"><div role="alert"><h2 class="govuk-error-summary__title"> There is a problem</h2><div class="govuk-error-summary__body"><ul class="govuk-list govuk-error-summary__list"><li><a href="#is-infrastructure-levy-formally-adopted-radio">Select whether levy formally adopted</a></li></ul></div></div></div>"`;
